File: kotlin/src/test/kotlin/com/joshy/hywrapper/RateLimitRetryTest.kt
package com.joshy.hywrapper

import kotlinx.coroutines.runBlocking
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class RateLimitRetryTest {
    private lateinit var server: MockWebServer

    @BeforeEach
    fun setup() {
        server = MockWebServer()
        server.start()
    }

    @AfterEach
    fun tearDown() {
        server.shutdown()
    }

    @Test
    fun `test auto retry on 429`() = runBlocking {
        val client = HypixelClient(
            "test-api-key",
            baseUrl = server.url("/").toString().removeSuffix("/"),
            autoRetry = true,
            maxRetries = 2
        )

        val rateLimitResponse = MockResponse()
            .setResponseCode(429)
            .setBody("""{"success": false, "cause": "Rate limit reached"}""")
            .addHeader("RateLimit-Limit", "300")
            .addHeader("RateLimit-Remaining", "0")
            .addHeader("RateLimit-Reset", "1")

        val successResponse = MockResponse()
            .setResponseCode(200)
            .setBody("""{"success": true, "name": "Bingo"}""")

        server.enqueue(rateLimitResponse)
        server.enqueue(successResponse)

        val response = client.getBingo()
        assertEquals(true, response.success)
        assertEquals("Bingo", response.name)
        assertEquals(2, server.requestCount)
    }

    @Test
    fun `test auto retry with Retry-After header`() = runBlocking {
        val client = HypixelClient(
            "test-api-key",
            baseUrl = server.url("/").toString().removeSuffix("/"),
            autoRetry = true,
            maxRetries = 2
        )

        val rateLimitResponse = MockResponse()
            .setResponseCode(429)
            .setBody("""{"success": false, "cause": "Rate limit reached"}""")
            .addHeader("Retry-After", "1")

        val successResponse = MockResponse()
            .setResponseCode(200)
            .setBody("""{"success": true, "name": "Bingo"}""")

        server.enqueue(rateLimitResponse)
        server.enqueue(successResponse)

        val response = client.getBingo()
        assertEquals(true, response.success)
        assertEquals(2, server.requestCount)
    }

    @Test
    fun `test max retries exceeded`() = runBlocking {
        val client = HypixelClient(
            "test-api-key",
            baseUrl = server.url("/").toString().removeSuffix("/"),
            autoRetry = true,
            maxRetries = 1
        )

        val rateLimitResponse = MockResponse()
            .setResponseCode(429)
            .setBody("""{"success": false, "cause": "Rate limit reached"}""")
            .addHeader("RateLimit-Reset", "1")

        server.enqueue(rateLimitResponse)
        server.enqueue(rateLimitResponse)
        server.enqueue(rateLimitResponse)

        assertFailsWith<RateLimitException> {
            client.getBingo()
        }
        assertEquals(2, server.requestCount) // Initial call + 1 retry
    }

    @Test
    fun `test no retry when autoRetry is false`() = runBlocking {
        val client = HypixelClient(
            "test-api-key",
            baseUrl = server.url("/").toString().removeSuffix("/"),
            autoRetry = false
        )

        val rateLimitResponse = MockResponse()
            .setResponseCode(429)
            .setBody("""{"success": false, "cause": "Rate limit reached"}""")

        server.enqueue(rateLimitResponse)

        assertFailsWith<RateLimitException> {
            client.getBingo()
        }
        assertEquals(1, server.requestCount)
    }
}

File: typescript/tests/rate_limit_retry.test.ts
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import { HypixelClient } from "../src/client";
import { RateLimitException } from "../src/types";

describe("HypixelClient Rate Limit Retry", () => {
    let mock: MockAdapter;
    const apiKey = "test-api-key";

    beforeEach(() => {
        mock = new MockAdapter(axios);
    });

    afterEach(() => {
        mock.restore();
    });

    test("should auto retry on 429", async () => {
        const client = new HypixelClient({
            api_key: apiKey,
            auto_retry: true,
            max_retries: 2
        });

        mock.onGet("/skyblock/bingo").replyOnce(429,
            { success: false, cause: "Rate limit reached" },
            {
                "ratelimit-limit": "300",
                "ratelimit-remaining": "0",
                "ratelimit-reset": "0"
            }
        );

        mock.onGet("/skyblock/bingo").replyOnce(200, { success: true, name: "Bingo" });

        const response = await client.getBingo();
        expect(response.success).toBe(true);
        expect(response.name).toBe("Bingo");
        expect(mock.history.get.length).toBe(2);
    });

    test("should auto retry with retry-after header", async () => {
        const client = new HypixelClient({
            api_key: apiKey,
            auto_retry: true,
            max_retries: 2
        });

        mock.onGet("/skyblock/bingo").replyOnce(429, 
            { success: false, cause: "Rate limit reached" },
            { "retry-after": "0" }
        );

        mock.onGet("/skyblock/bingo").replyOnce(200, { success: true, name: "Bingo" });

        const response = await client.getBingo();
        expect(response.success).toBe(true);
        expect(mock.history.get.length).toBe(2);
    });

    test("should fail after max retries", async () => {
        const client = new HypixelClient({
            api_key: apiKey,
            auto_retry: true,
            max_retries: 1
        });

        mock.onGet("/skyblock/bingo").reply(429, 
            { success: false, cause: "Rate limit reached" },
            { "ratelimit-reset": "0" }
        );

        await expect(client.getBingo()).rejects.toThrow(RateLimitException);
        expect(mock.history.get.length).toBe(2);
    });

    test("should not retry when disabled", async () => {
        const client = new HypixelClient({
            api_key: apiKey,
            auto_retry: false
        });

        mock.onGet("/skyblock/bingo").reply(429, 
            { success: false, cause: "Rate limit reached" }
        );

        await expect(client.getBingo()).rejects.toThrow(RateLimitException);
        expect(mock.history.get.length).toBe(1);
    });
});
